apps/desktop-wallet/src/contexts/global.tsx

```diff
@@ -82,58 +82,85 @@ export const GlobalContextProvider: FC<{ overrideContextValue?: PartialDeep<Glob
   const triggerNewVersionDownload = () => setNewVersionDownloadTriggered(true)
   const resetNewVersionDownloadTrigger = () => setNewVersionDownloadTriggered(false)
 
-  const unlockWallet = async ({ event, walletId, password, afterUnlock, passphrase }: WalletUnlockProps) => {
-    let wallet
+  const unlockWallet = async (props: WalletUnlockProps | null) => {
+    if (!props) return
+
+    const { event, walletId, afterUnlock } = props
+    let { password, passphrase } = props
+    let encryptedWallet: StoredEncryptedWallet | null
+    let initialAddress: NonSensitiveAddressData
+    let version: number
 
     try {
-      wallet = WalletStorage.load(walletId, password)
+      encryptedWallet = WalletStorage.load(walletId)
+
+      version = keyring.decryptAndCacheMnemonicAndPassword(encryptedWallet.encrypted, password, passphrase ?? '')
     } catch (e) {
       console.error(e)
       dispatch(passwordValidationFailed())
       return
     }
 
-    if (passphrase) {
-      wallet = { ...wallet, ...getWalletFromMnemonic(wallet.mnemonic, passphrase) }
+    // In version 1 the encrypted mnemonic used to be stored as a string before we started using Buffer. This migrates
+    // the encrypted wallet from StoredStateV1 to StoredStateV2.
+    if (version === 1) {
+      try {
+        WalletStorage.update(walletId, { encrypted: keyring.encryptMnemonicForStorageAndCachePassword(password) })
+        console.log('Migrated stored mnemonic from version 1 to version 2')
+      } catch (e) {
+        console.error('Failed migrating stored mnemonic from version 1 to version 2', e)
+      }
+    }
```

Comment:
I purposefully didn't want to create a migration script for this one because I want to avoid passing the password around.